# Начнем с мемов

![title](memes/meme_1.jpg) 
![title](memes/meme_2.jpg) 
![title](memes/meme_3.jpg) 
![title](memes/meme_4.jpg) 

# Data mining with Python

#### Обычно очень много внимания в процессе обучения анализу данных уделяется алгоритмам и методам обработки данных и намного меньше внимания уделяется тому эти данные взять. При этом: <br>
- Без самих данных навряд ли что-то получится; <br>
- От качества данных зависит результат работы любого алгоритма; (часто бывает важнее алгоритмов)<br>
- Часто работают лучше те модели которые обучались на более качественных или большем объеме данных (70-80% работа с данными, 20-30% пробования алгоритмов).

### Где брать данные?

Конечно, данные вам могут предоставить на работе или заказчик. Вы можете работать в одной из компаний где данных больше чем предостаточно. А что если их не достаточно или в них нет того что вам нужно.


### В этом случае лучше всего действовать в следующей последовательности:

- Найти готовый датасет;
- Собрать данные самому:
   - для этого можно воспользоваться API;
   - собрать данные со страниц сайтов;
   - нагенерить/создать (видео снимали с датасетом) http://techno-start.ru/project/241/
   
   Существует несколько инструментов для парсинга данных самостоятельно но чаще всего используют:
    - библиотеку requests + библиотеку beautiful soup
    - фреймворк Scrappy 
- Купить (данные новая нефть, иногда бывает выгоднее, чем потраить время фултайм программиста на скрапинг). Что делать стартапам? https://www.opendota.com/


### Поиск готового датасета.
Логичнее всего начать с поиска готового датасета. Существует много сайтов где можно найти уже готовые данные, например, очень много данных есть на платформе <a href="https://www.kaggle.com/datasets">Kaggle<a/>. Менее популярная платформа, но не менее интересная <a href="http://academictorrents.com/browse.php">academictorrents.com </a>, https://datasetsearch.research.google.com/ советую проверить. <br>
 
ODS Community datasets https://ods.ai/<br>
    
Существуют и еще менее тривиальные источники данных о существовании которых иногда трудно догодаться. Так например, есть Google Big Querry на котором сейчас выложено около 40 публичных датасетов, один из них это все записи форума reddit. Казалось бы зачем нам форум такой форум? На самом деле применений может быть очень много, например, по его данным легко обучить классификатор текстов на разные категории, а категорий на форуме очень много. Или случай со спором<br>
Интерфейс у Google Big Querry это SQL, что очень удобно и можно сразу вытащить то что нужно. Попробовать можно здесь: <a href="https://bigquery.cloud.google.com/results/bigdatagame-200413:US.bquijob_609efa7_16871e6cd5e?pli=1">Google Big Query</a> <br>
Пример запроса который достает 1500 комментариве из раздела "Art" c лайками больше 5ти. <br>
    
<strong><i>SELECT body, score FROM <br>
`fh-bigquery.reddit_comments.2017_05` <br>
WHERE subreddit LIKE 'Art' AND score > 5 <br>
LIMIT 15000 <br>
</i></strong>

In [ ]:
### Самостоятельный сбор данных с помощью API.
Ок, если не удалось найти готовый датасет, то возможно они есть в отрытом доступе на каком-то сайте или платформе через общедоступный API. <br>
 <br>
#### Работа с API выглядит следующим образом. 
Есть два метода (на самом деле их больше GET, OPTIONS, POST, PUT, PATCH or DELETE) post и get с помощью которых мы можем обратиться к сайту или платформе. С помощью метода get мы можем можем сделать запрос ( напрмер когда мы открываем сайт по ссылке то это get), а с помощью метода post мы можем передать какеи-то значения клиенту.<br>
Так например, сайт HH.ru предоставляет свой API и по следующему запросу мы можем получить список вакансий определенной категории 1.221 (цифры обозначающие вакансии это их внутренняя кухня, тут без лишних деталей). <br>
https://api.hh.ru/vacancies?specialization=1.221 <br>
В ответ на странице вы увидите скорее всего Json фаил, который очень легко читать и с которым приятно работать. <br>
Теперь нам нужно понять что с этим делать дальше.
Минусы? (зачем)

### Работа с библиотекой requests
Собственно для работы с post и get запросами нам и понадобиться библиотека requests.

Для установки библиотеки используйте следующую комманду в терминале (если на Linux)

pip install requests

Документация по библиотеке requests: https://requests.readthedocs.io/en/master/

### Посмотрим на примере сайта HH.ru как работать с библиотекой и как можно получить какие-то данные с помощью API

In [17]:
import requests
url1 = 'https://api.hh.ru/vacancies?specialization=1.221'
url2 = 'https://api.hh.ru/vacancies/1'
resp = requests.get(url2)
print(resp.status_code)
resp = requests.get(url1)
print(resp.status_code)

404
200


#### Мы видим что для одной из ссылок клиент вернул статус 404, а для другой 200.

Это код ответа от клиента, код 200 обозначает что все ок, а код 404 обозначает Not found. Полный списко кодов состояния HTTP можно посмотреть <a href="https://ru.wikipedia.org/wiki/Список_кодов_состояния_HTTP"> вот здесь. </a> В дальнейшем, в рамках этого туториала нас интересует только код 200. Но на практике полезно знать и другие. (Для обработки исключений, ФБК)<br>
И так, для одной из ссылок мы получили 200, все ок, двигаемся дальше.

In [18]:
print(resp.text)
print(type(resp.text))

{"items":[{"id":"36147933","premium":false,"name":"Web-дизайнер","department":null,"has_test":false,"response_letter_required":false,"area":{"id":"1","name":"Москва","url":"https://api.hh.ru/areas/1"},"salary":{"from":80000,"to":100000,"currency":"RUR","gross":false},"type":{"id":"open","name":"Открытая"},"address":{"city":"Москва","street":"Дегтярный переулок","building":"5с2","description":null,"lat":55.768213,"lng":37.601331,"raw":"Москва, Дегтярный переулок, 5с2","metro":{"station_name":"Тверская","line_name":"Замоскворецкая","station_id":"2.122","line_id":"2","lat":55.765343,"lng":37.603918},"metro_stations":[{"station_name":"Тверская","line_name":"Замоскворецкая","station_id":"2.122","line_id":"2","lat":55.765343,"lng":37.603918},{"station_name":"Пушкинская","line_name":"Таганско-Краснопресненская","station_id":"7.124","line_id":"7","lat":55.765607,"lng":37.604356},{"station_name":"Маяковская","line_name":"Замоскворецкая","station_id":"2.80","line_id":"2","lat":55.769808,"lng":37

In [28]:
resp.url

'https://api.hh.ru/vacancies?specialization=1.221&page=2&per_page200'

In [30]:
resp.encoding

'UTF-8'

In [19]:
resp.headers

{'Server': 'nginx', 'Date': 'Thu, 19 Mar 2020 21:37:45 GMT', 'Content-Type': 'application/json; charset=UTF-8', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Keep-Alive': 'timeout=60', 'X-Request-Id': '158465386561760b769a3f5642167269, 158465386561760b769a3f5642167269', 'Cache-Control': 'max-age=0, private, must-revalidate', 'Expires': 'Thu, 06 Feb 2020 21:37:45 GMT', 'Access-Control-Allow-Origin': '*', 'Access-Control-Expose-Headers': 'Location, ETag, Date, Expires, Cache-Control, Content-Type, X-Request-ID', 'X-Frame-Options': 'SAMEORIGIN', 'X-Content-Type-Options': 'nosniff', 'Content-Encoding': 'gzip'}

In [21]:
resp.json()

{'items': [{'id': '36147933',
   'premium': False,
   'name': 'Web-дизайнер',
   'department': None,
   'has_test': False,
   'response_letter_required': False,
   'area': {'id': '1', 'name': 'Москва', 'url': 'https://api.hh.ru/areas/1'},
   'salary': {'from': 80000, 'to': 100000, 'currency': 'RUR', 'gross': False},
   'type': {'id': 'open', 'name': 'Открытая'},
   'address': {'city': 'Москва',
    'street': 'Дегтярный переулок',
    'building': '5с2',
    'description': None,
    'lat': 55.768213,
    'lng': 37.601331,
    'raw': 'Москва, Дегтярный переулок, 5с2',
    'metro': {'station_name': 'Тверская',
     'line_name': 'Замоскворецкая',
     'station_id': '2.122',
     'line_id': '2',
     'lat': 55.765343,
     'lng': 37.603918},
    'metro_stations': [{'station_name': 'Тверская',
      'line_name': 'Замоскворецкая',
      'station_id': '2.122',
      'line_id': '2',
      'lat': 55.765343,
      'lng': 37.603918},
     {'station_name': 'Пушкинская',
      'line_name': 'Таганско-

In [26]:
url=url1+'&page=2&per_page200'
resp = requests.get(url)
print(resp)

<Response [200]>


'&page=1&per_page100'  - это опять внутренняя кухня API HH.ru, у всех платформ она разная. Здесь мы говорим что мы загружаем первую страницу и 100 результатов на странице. Почему нельзя сделать больше или подругому за раз зависит от того как это решили разработчики API. <br>
Но можно получить больше результатов следующим образом.


In [105]:
searches = []
for page in range(20): #что если больше?
    searches.append(
        requests.get(url1 + '&page={}&per_page=100'.format(page)).json()
    )

In [68]:
#Можно открыть первую ссылку в нашем дасатае и посмотреть что мы собрали.
searches[0]['items'][0]['url']

'https://api.hh.ru/vacancies/36147933?host=hh.ru'

In [69]:
searches[0]['items'][0].keys()

dict_keys(['id', 'premium', 'name', 'department', 'has_test', 'response_letter_required', 'area', 'salary', 'type', 'address', 'response_url', 'sort_point_distance', 'employer', 'published_at', 'created_at', 'archived', 'apply_alternate_url', 'insider_interview', 'url', 'alternate_url', 'relations', 'snippet', 'contacts'])

In [70]:
vacs = [x['items'] for x in searches]
vacs = sum(list(vacs),[])
vacs[0]

{'id': '36147933',
 'premium': False,
 'name': 'Web-дизайнер',
 'department': None,
 'has_test': False,
 'response_letter_required': False,
 'area': {'id': '1', 'name': 'Москва', 'url': 'https://api.hh.ru/areas/1'},
 'salary': {'from': 80000, 'to': 100000, 'currency': 'RUR', 'gross': False},
 'type': {'id': 'open', 'name': 'Открытая'},
 'address': {'city': 'Москва',
  'street': 'Дегтярный переулок',
  'building': '5с2',
  'description': None,
  'lat': 55.768213,
  'lng': 37.601331,
  'raw': 'Москва, Дегтярный переулок, 5с2',
  'metro': {'station_name': 'Тверская',
   'line_name': 'Замоскворецкая',
   'station_id': '2.122',
   'line_id': '2',
   'lat': 55.765343,
   'lng': 37.603918},
  'metro_stations': [{'station_name': 'Тверская',
    'line_name': 'Замоскворецкая',
    'station_id': '2.122',
    'line_id': '2',
    'lat': 55.765343,
    'lng': 37.603918},
   {'station_name': 'Пушкинская',
    'line_name': 'Таганско-Краснопресненская',
    'station_id': '7.124',
    'line_id': '7',
  

In [71]:
#Посморим на первые 5 названий вакансий
titles = [item['name'].lower() for item in vacs]
titles[:5]

['web-дизайнер',
 'дизайнер по доработке выпускных альбомов',
 'директор it-управления',
 'менеджер интернет-проектов / project manager',
 'дизайнер по доработке выпускных альбомов']

In [72]:
#А теперь очень просто найдем самое часто встречаемое название вакансии

max(set(titles), key=titles.count)

'программист 1с'

In [73]:
from collections import Counter
Counter(titles)

Counter({'web-дизайнер': 33,
         'дизайнер по доработке выпускных альбомов': 5,
         'директор it-управления': 1,
         'менеджер интернет-проектов / project manager': 1,
         'младший руководитель проектов': 1,
         'программист 1с (2ая линия поддержки) [id15849]': 1,
         'руководитель it проектов': 2,
         'ведущий web-разработчик php/js (full stack)': 1,
         'frontend developer (javascript, react)': 1,
         'product owner': 8,
         'руководитель проектов': 7,
         'дизайнер ui/ux': 2,
         'ведущий специалист поддержки систем 1с:еrp/ух [id21689/id19761]': 1,
         'дизайнер ux/ui (проектировщик интерфейсов)': 1,
         'разработчик java': 5,
         'дизайнер ux/ui/web/mobile': 2,
         'frontend-разработчик (react)': 9,
         'web-дизайнер креативщик': 1,
         'html-верстальщик / frontend программист': 1,
         'интернет-маркетолог/сeo специалист': 1,
         'руководитель it проекта в сфере жкх (разработка)': 1,

In [64]:
import re

REGEXP_1 = re.compile(r'(.*?)scientist(.*?)')
REGEXP_2 = re.compile(r'(.*?)data(.*?)')
for vacancy in vacs:
    if REGEXP_1.fullmatch(vacancy['name'].lower()) or REGEXP_2.fullmatch(vacancy['name'].lower()):
        print(vacancy['name'])
        

Data-инженер/ разработчик DWH
Продуктовый аналитик Big Data
Product owner финтех (Big Data)
Data Scientist
Data Engineer (SQL)
Преподаватель Python & Data Science
Junior Data Analyst
Data scientist/ Директор в управлении розничного моделирования
Senior Data Engineer (Teradata/GreenPlum).
Инженер-программист data science, computer vision (middle)
Data Engineer
Data Engineer
Frontend-разработчик (проекты Data Science)
Data engineer/DWH (Рекламные технологии)
Data scientist
Data engineer (ПРК)
Разработчик платформы рекомендательных систем. Data Engineer.
ETL-BigData Developer
Разработчик / Data Engineer (Big Data)
Стажер/Младший разработчик (DWH/ETL, Big data)
Middle Data Science в R&D
Data engineer (Отдел хранения и обработки данных)
Junior Data Scientist (рек.системы)
Data analyst
Data scientist
Senior Data Scientist
Data Scientist
Специалист-исследователь (Data science & information security)
Data Scientist in Oil & Gas
Senior Data Scientist
Python Developer (Data Scientist)
Senior Dat

### Посмотрим на зарплаты, как обещали

In [100]:
REGEXP_1 = re.compile(r'(.*?)scientist(.*?)')
REGEXP_2 = re.compile(r'(.*?)data(.*?)')
lower_bounds = []
upper_bounds = []
good_vacs = [vacancy for vacancy in vacs if REGEXP_1.fullmatch(vacancy['name'].lower()) or REGEXP_2.fullmatch(vacancy['name'].lower())]
for vacancy in good_vacs:
    if 'salary' in vacancy:
        print(vacancy['salary'])

None
None
None
None
None
{'from': 150000, 'to': None, 'currency': 'RUR', 'gross': True}
None
None
None
None
None
None
None
None
None
None
None
None
None
None
{'from': 130000, 'to': 180000, 'currency': 'RUR', 'gross': True}
None
None
None
None
None
None
None
None
None
{'from': 3000, 'to': 3200, 'currency': 'USD', 'gross': True}
None


### ЖИЗНЬ БОЛЬ

Использовать API можно не только для текстовой информации, хотя чаще всего для нее. Вот пример работы c API Flickr https://gist.github.com/diff7/627a306d6d8e4ecd0162808deae7d244 для получения картинок. Для использования нужно импортировать API key на сайте и установить пакеты, которые импортируются в начале скрипта, если они еще не стоят. <br>
https://vk.com/dev/methods - vk

А если вам вдруг понравилось работать с апишками то вот здесь есть очень хороший проект с кучей всяких разных API со всего интернета.
<a href="https://public-apis.xyz/">public-apis.xyz</a>

In [90]:
tok = 'b2541f06b2541f06b2541f063fb21ab6a8bb254b2541f06ec27e6299445529a3a6188f2'
url = 'https://api.vk.com/method/users.get?user_id=210700286&v=5.52&access_token={}'.format(tok)
resp = requests.get(url)

In [91]:
resp.text

'{"response":[{"id":210700286,"first_name":"Lindsey","last_name":"Stirling"}]}'

###  Beautiful Soup
Документация : https://www.crummy.com/software/BeautifulSoup/bs4/doc/ <br>

Чаще всего через API можно получить намного больше интересной и качественной информации, убедиться в этом можно опять же зайдя на сайт public-apis.xyz там есть и статистика из онлайн игр отзывы по фильмам и много всего. Но  иногда сайты не предоставляют API, но данные на них все таки есть. (Мы не обсуждаем можно ли их использовать или нет, скорее всего об этом написано на самом сайте) Тогда приходится дейстовать подругому. <br>
Посмотрим на примере сайта http://quotes.toscrape.com который был создан для как раз для практики данных навыков. <br>

In [110]:
link='http://quotes.toscrape.com'
page=requests.get(link)
if page.status_code == 200:  
    print(page.text)

<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<title>Quotes to Scrape</title>
    <link rel="stylesheet" href="/static/bootstrap.min.css">
    <link rel="stylesheet" href="/static/main.css">
</head>
<body>
    <div class="container">
        <div class="row header-box">
            <div class="col-md-8">
                <h1>
                    <a href="/" style="text-decoration: none">Quotes to Scrape</a>
                </h1>
            </div>
            <div class="col-md-4">
                <p>
                
                    <a href="/login">Login</a>
                
                </p>
            </div>
        </div>
    

<div class="row">
    <div class="col-md-8">

    <div class="quote" itemscope itemtype="http://schema.org/CreativeWork">
        <span class="text" itemprop="text">“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”</span>
        <span>by <small class="author" itempr

Как видно здесь присутствует какой - то текст, но ответ от страницы нам пришел так как видит ее браузер, т.е. с HTML тегами, сss и возможно JavaScript. Собственно что бы очистить страницу от HTML и всего прочего нам и нужна библиотека Beautiful Soup.

### Чуть - чуть про HTML, если вы в курсе то пропустите.

<!DOCTYPE html>  
<html>  
    <head  style="color:red">Это простой пример HTML разметки используемый для форматирования текста.</head>
    <div> Если нажаит на <strong>этот текст</strong> два раза, то вы увидите его так как его видит браузер.</div>
    <body>
        <h1><span class="cm-string"> Заголовок &lt;h1&gt; </span></h1>
        <p>Параграф &lt;p&gt; </p>
        <a href="www.nowhere.space">Ссылка в никуда</a>
        <div>блок &lt;div&gt; </div>
        <h2  style="color:red" > И конечно картинка: </h2>
        <img src="https://cs6.pikabu.ru/post_img/big/2015/06/18/3/1434596941_632146314.jpg">
        <p style="color:blue;">Для того что бы просмотреть страницу в HTML можно нажать правой кнопкой мыши и выбрать "inspect" в появившимся меню.</p>
        <p>В общем, функционал зависит от браузера, но в целом должно появится что-то вроде этого. Можете попробовать с сайтом
            <a href="http://quotes.toscrape.com"> http://quotes.toscrape.com </a></p>
        <img src="https://mydatacareer.com/wp-content/uploads/2017/05/scraping3.png">
    </body>
</html>

#### Теперь мы знаем каким элементом страницы является тот или иной текст и сможем легко его извлечь с помощью библиотеки Beautiful Soup

Для установки библиотеки используйте следующую комманду в терминале (если на Linux)

pip install beautifulsoup4

In [111]:
import bs4 as bs

link='http://quotes.toscrape.com'
page=requests.get(link)
if page.status_code == 200:  
    soup = bs.BeautifulSoup(page.content, 'html.parser')
    
soup.find_all('span', attrs={'class': 'text'}) #Как видно из картинки выше, наша цитата находится внутри тега span c классом text

[<span class="text" itemprop="text">“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”</span>,
 <span class="text" itemprop="text">“It is our choices, Harry, that show what we truly are, far more than our abilities.”</span>,
 <span class="text" itemprop="text">“There are only two ways to live your life. One is as though nothing is a miracle. The other is as though everything is a miracle.”</span>,
 <span class="text" itemprop="text">“The person, be it gentleman or lady, who has not pleasure in a good novel, must be intolerably stupid.”</span>,
 <span class="text" itemprop="text">“Imperfection is beauty, madness is genius and it's better to be absolutely ridiculous than absolutely boring.”</span>,
 <span class="text" itemprop="text">“Try not to become a man of success. Rather become a man of value.”</span>,
 <span class="text" itemprop="text">“It is better to be hated for what you are than to be loved for what you are not.

In [112]:
tags=soup.find_all('span', attrs={'class': 'text'})
tags[0].text

'“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”'

In [113]:
[item.text for item in tags]

['“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”',
 '“It is our choices, Harry, that show what we truly are, far more than our abilities.”',
 '“There are only two ways to live your life. One is as though nothing is a miracle. The other is as though everything is a miracle.”',
 '“The person, be it gentleman or lady, who has not pleasure in a good novel, must be intolerably stupid.”',
 "“Imperfection is beauty, madness is genius and it's better to be absolutely ridiculous than absolutely boring.”",
 '“Try not to become a man of success. Rather become a man of value.”',
 '“It is better to be hated for what you are than to be loved for what you are not.”',
 "“I have not failed. I've just found 10,000 ways that won't work.”",
 "“A woman is like a tea bag; you never know how strong it is until it's in hot water.”",
 '“A day without sunshine is like, you know, night.”']

### Недостатки такого метода

С помощью комбинации Beautiful Soup + Requests можно получить данные с большого числа сайтов. Но существует ряд недостатков.

- Достаточно долго, так как процесс не параллелиться;
- Для каждого нового сайта приходится писать все заново;
- Некоторые сайты блокируют возможность парсинга, а некоторые используют JavaScript для отрисовки вместо HTML, тогда связки Beautiful Soup + Requests будет недостаточно.

### Scrapy
Scrapy это уже не библиотека, а полноценный фреймворк для парсинга данных с сайтов с кучей настроек и шикарным функционалом.
Основным недостатком Scrapy является достаточно высокий порог входа, придется потратить некоторое время что бы разобраться во всех деталях. По дизайну фреймворк часто сравнивают с Django.

Scrapy лучше использовать в следующих случаях:
- Когда важна скорость скачивания;
- Когда данные парсятся с большого числа однотипных сайтов;
- Когда предстоит парсить данные с одого или нескольких сайтов на регулярной основе;
- Когда нужно спарсить очень большой сайт.

### Работа со Scrapy на примере сайте quotes.toscrape.com

В командной строке

pip install scrapy

дальнейшие команды нужно делать в системном питоне

In [ ]:
#запустить Scrapy проект 
scrapy startproject proj

#в директории проекта должна появится папка spiders
cd scrapy/proj/proj/spiders 

#cоздайте в директории spiders фаил quotes_spider.py
nano quotes_spider.py

In [17]:
#содержимым файла будет соледующий код

# -*- coding: utf-8 -*-
import scrapy

class QuotesSpiderSpider(scrapy.Spider):
    name = 'quotes_spider' #название паука
    allowed_domains = ['quotes.toscrape.com'] 
    start_urls = ['http://quotes.toscrape.com/'] #ссылка с которой начинается парсинг

    def parse(self, response):
        #для выделения HTML тегов или сss декораторов Scrapy использует Xpath
        #по сути здесь мы передаем теже параметры которые передавали в Beautiful Soup
        quotes = response.xpath("//div[@class='quote']//span[@class='text']/text()").extract()
        yield {'quotes': quotes}

scrapy runspider quotes_spider.py -o quotes.json